### PR TITLE
Material Desktop Context Menu

### LINKED ISSUE #31955: Should implement a context menu widget
With the ability to right-click, we need to be able to allow developers to implement custom context menus.  On desktop platforms, these may need to be platform-specific context menus (so that they can render outside of the window), on web they will need to integrate with the context menu APIs there, and on Android they will need to be rendered by Flutter (one can right click on ChromeOS in an Android app, for instance. iOS doesn't support mouse, so it's not needed there). As a first pass on desktop platforms, it may initially be acceptable if the menu is rendered by Flutter (and hence doesn't render outside of the window), but long term probably not, so the API should definitely take that into account (and not take Widgets for the items, but rather some form of data object).

Before this, #31946 will probably need to be implemented, in order to provide the actions that the menu will invoke.

See also #30976 and #31412

### PR DESCRIPTION
**Pull request by justinmc**
A desktop right-click menu for Windows and Linux, in a Material-esque style.

| Light | Dark |
| ----- | ----- |
| Screen Shot 2021-01-20 at 9 59 16 AM | Screen Shot 2021-01-20 at 9 59 35 AM |

~~This is branched off of https://github.com/flutter/flutter/pull/73882 and needs to wait for that to be merged first.~~

See the list of known limitations tracked in https://github.com/flutter/flutter/issues/74255.

Closes https://github.com/flutter/flutter/issues/31955

### CODE REVIEW

File: packages/flutter/lib/src/cupertino/desktop_text_selection.dart [hunk 1]
```diff
@@ -319,48 +319,6 @@ class _CupertinoDesktopTextSelectionToolbar extends StatelessWidget {
...
 
-// Positions the toolbar at [anchor] if it fits, otherwise moves it so that it
-// just fits fully on-screen.
-//
-// See also:
-//
-//   * [CupertinoDesktopTextSelectionToolbar], which uses this to position itself.
-//   * [TextSelectionToolbarLayoutDelegate], which does a similar layout for
-//     the mobile text selection toolbars.
-class _DesktopTextSelectionToolbarLayoutDelegate extends SingleChildLayoutDelegate {
```
Comment (justinmc) [inline packages/flutter/lib/src/cupertino/desktop_text_selection.dart:330]: I moved this into its own file and made it public in order to share it with the Mac and Material desktop context menus.

**Review by HansMuller — APPROVED** [review]
LGTM

**Comment by creativecreatorormaybenot** [comment]
I am still confused as to how to use this. @justinmc are there full examples explaining usage? I have never seen this in a Flutter app. [From the docs](https://api.flutter.dev/flutter/material/TextField/selectionControls.html), I am not sure how to use it. 

> TextField, a Material Design themed wrapper of EditableText, which shows the selection toolbar upon appropriate user events based on the user's platform set in ThemeData.platform.

It sounds like it should appear on right click by default but it does not.

**Comment by justinmc** [comment]
This should show up by default on desktop, but only when right clicking on EditableText-based widgets.  Right clicking elsewhere does nothing, for now.

This is also purposely not customizable or reusable for now.  It was a temporary solution that we plan to expand on.

**Comment by creativecreatorormaybenot** [comment]
@justinmc Thanks. Does the first part mean that it does not show up on web?

**Comment by justinmc** [comment]
Yeah that's right.  Web will still show the browser's right click menu.

**Pull request merged**